### ISSUE TITLE
[Phase 5] Gradient Boosting Ensemble

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/demand-forecasting-engine#6]
## Summary

Combine LightGBM, XGBoost, and CatBoost into a weighted ensemble for improved forecast accuracy and robustness. Different gradient boosting implementations have different inductive biases that can complement each other.

## System Context

```
demand-forecasting-engine/
├── src/
│   └── models/
│       ├── base_model.py           # EXISTS
│       ├── lightgbm_model.py       # EXISTS
│       ├── quantile_model.py       # EXISTS
│       ├── xgboost_model.py        # <- CREATE
│       ├── catboost_model.py       # <- CREATE
│       └── ensemble.py             # <- CREATE
└── scripts/
    └── train_ensemble.py           # <- CREATE
```

## Dependencies

- Requires: Phase 4 (Quantile Regression)
- Blocks: Phase 6 (Hierarchical Reconciliation)

### 1. XGBoost Model (`src/models/xgboost_model.py`)

```python
import numpy as np
import pandas as pd
import xgboost as xgb
from typing import Optional, Dict, Any, List
from .base_model import BaseForecaster


class XGBoostForecaster(BaseForecaster):
    """
    XGBoost-based demand forecaster.

    Parameters match LightGBMForecaster for consistency.
    """

    def __init__(
        self,
        n_estimators: int = 1000,
        learning_rate: float = 0.05,
        max_depth: int = 6,
        min_child_weight: int = 1,
        subsample: float = 0.8,
        colsample_bytree: float = 0.8,
        reg_alpha: float = 0.1,
        reg_lambda: float = 0.1,
        random_state: int = 42,
        n_jobs: int = -1,
        early_stopping_rounds: int = 50,
        verbose: int = 0,
    ):
        self.n_estimators = n_estimators
        self.learning_rate = learning_rate
        self.max_depth = max_depth
        self.min_child_weight = min_child_weight
        self.subsample = subsample
        self.colsample_bytree = colsample_bytree
        self.reg_alpha = reg_alpha
        self.reg_lambda = reg_lambda
        self.random_state = random_state
        self.n_jobs = n_jobs
        self.early_stopping_rounds = early_stopping_rounds
        self.verbose = verbose

        self.model_: Optional[xgb.Booster] = None
        self.feature_names_: Optional[List[str]] = None

    def fit(
        self,
        X: pd.DataFrame,
        y: pd.Series,
        eval_set: Optional[tuple] = None,
    ) -> 'XGBoostForecaster':
        """Train XGBoost model."""
        self.feature_names_ = list(X.columns)

        dtrain = xgb.DMatrix(X, label=y, feature_names=self.feature_names_)

        evals = [(dtrain, 'train')]
        if eval_set is not None:
            X_val, y_val = eval_set
            dval = xgb.DMatrix(X_val, label=y_val, feature_names=self.feature_names_)
            evals.append((dval, 'valid'))

        params = {
            'objective': 'reg:squarederror',
            'eval_metric': 'rmse',
            'learning_rate': self.learning_rate,
            'max_depth': self.max_depth,
            'min_child_weight': self.min_child_weight,
            'subsample': self.subsample,
            'colsample_bytree': self.colsample_bytree,
            'reg_alpha': self.reg_alpha,
            'reg_lambda': self.reg_lambda,
            'seed': self.random_state,
            'nthread': self.n_jobs,
            'verbosity': self.verbose,
        }

        self.model_ = xgb.train(
            params,
            dtrain,
            num_boost_round=self.n_estimators,
            evals=evals,
            early_stopping_rounds=self.early_stopping_rounds if eval_set else None,
            verbose_eval=self.verbose > 0,
        )

        return self

    def predict(self, X: pd.DataFrame) -> np.ndarray:
        """Generate predictions."""
        dtest = xgb.DMatrix(X, feature_names=self.feature_names_)
        return self.model_.predict(dtest)

    def get_feature_importance(self) -> pd.DataFrame:
        """Get feature importance."""
        importance = self.model_.get_score(importance_type='gain')
        return pd.DataFrame({
            'feature': list(importance.keys()),
            'importance': list(importance.values())
        }).sort_values('importance', ascending=False)

    def get_params(self) -> Dict[str, Any]:
        return {
            'n_estimators': self.n_estimators,
            'learning_rate': self.learning_rate,
            'max_depth': self.max_depth,
            'min_child_weight': self.min_child_weight,
            'subsample': self.subsample,
            'colsample_bytree': self.colsample_bytree,
            'reg_alpha': self.reg_alpha,
            'reg_lambda': self.reg_lambda,
        }

    def set_params(self, **params) -> 'XGBoostForecaster':
        for key, value in params.items():
            if hasattr(self, key):
                setattr(self, key, value)
        return self
```

### 2. CatBoost Model (`src/models/catboost_model.py`)

```python
import numpy as np
import pandas as pd
from catboost import CatBoostRegressor
from typing import Optional, Dict, Any, List
from .base_model import BaseForecaster


class CatBoostForecaster(BaseForecaster):
    """
    CatBoost-based demand forecaster.

    CatBoost handles categorical features natively without encoding.
    """

    def __init__(
        self,
        n_estimators: int = 1000,
        learning_rate: float = 0.05,
        max_depth: int = 6,
        l2_leaf_reg: float = 3.0,
        random_strength: float = 1.0,
        bagging_temperature: float = 1.0,
        random_state: int = 42,
        early_stopping_rounds: int = 50,
        verbose: int = 0,
        cat_features: Optional[List[str]] = None,
    ):
        self.n_estimators = n_estimators
        self.learning_rate = learning_rate
        self.max_depth = max_depth
        self.l2_leaf_reg = l2_leaf_reg
        self.random_strength = random_strength
        self.bagging_temperature = bagging_temperature
        self.random_state = random_state
        self.early_stopping_rounds = early_stopping_rounds
        self.verbose = verbose
        self.cat_features = cat_features

        self.model_: Optional[CatBoostRegressor] = None
        self.feature_names_: Optional[List[str]] = None

    def fit(
        self,
        X: pd.DataFrame,
        y: pd.Series,
        eval_set: Optional[tuple] = None,
    ) -> 'CatBoostForecaster':
        """Train CatBoost model."""
        self.feature_names_ = list(X.columns)

        # Auto-detect categorical columns if not specified
        if self.cat_features is None:
            self.cat_features = list(X.select_dtypes(include=['object', 'category']).columns)

        self.model_ = CatBoostRegressor(
            iterations=self.n_estimators,
            learning_rate=self.learning_rate,
            depth=self.max_depth,
            l2_leaf_reg=self.l2_leaf_reg,
            random_strength=self.random_strength,
            bagging_temperature=self.bagging_temperature,
            random_seed=self.random_state,
            early_stopping_rounds=self.early_stopping_rounds if eval_set else None,
            verbose=self.verbose,
            cat_features=self.cat_features if self.cat_features else None,
        )

        self.model_.fit(
            X, y,
            eval_set=eval_set if eval_set else None,
        )

        return self

    def predict(self, X: pd.DataFrame) -> np.ndarray:
        """Generate predictions."""
        return self.model_.predict(X)

    def get_feature_importance(self) -> pd.DataFrame:
        """Get feature importance."""
        importance = self.model_.get_feature_importance()
        return pd.DataFrame({
            'feature': self.feature_names_,
            'importance': importance
        }).sort_values('importance', ascending=False)

    def get_params(self) -> Dict[str, Any]:
        return {
            'n_estimators': self.n_estimators,
            'learning_rate': self.learning_rate,
            'max_depth': self.max_depth,
            'l2_leaf_reg': self.l2_leaf_reg,
            'random_strength': self.random_strength,
            'bagging_temperature': self.bagging_temperature,
        }

    def set_params(self, **params) -> 'CatBoostForecaster':
        for key, value in params.items():
            if hasattr(self, key):
                setattr(self, key, value)
        return self
```

### 3. Ensemble Model (`src/models/ensemble.py`)

```python
import numpy as np
import pandas as pd
from typing import Dict, List, Optional, Any, Union
from scipy.optimize import minimize

from .base_model import BaseForecaster
from .lightgbm_model import LightGBMForecaster
from .xgboost_model import XGBoostForecaster
from .catboost_model import CatBoostForecaster


MODEL_REGISTRY = {
    'lightgbm': LightGBMForecaster,
    'xgboost': XGBoostForecaster,
    'catboost': CatBoostForecaster,
}


class ForecastingEnsemble(BaseForecaster):
    """
    Ensemble of gradient boosting models.

    Parameters
    ----------
    models : list of str
        Model names to include ('lightgbm', 'xgboost', 'catboost')
    weights : str or list of float
        'equal': Equal weights
        'optimize': Learn weights from validation data
        list: Manual weights (must sum to 1)
    model_params : dict, optional
        Parameters per model: {'lightgbm': {...}, 'xgboost': {...}}
    """

    def __init__(
        self,
        models: List[str] = ['lightgbm', 'xgboost', 'catboost'],
        weights: Union[str, List[float]] = 'equal',
        model_params: Optional[Dict[str, Dict]] = None,
    ):
        self.model_names = models
        self.weights_strategy = weights
        self.model_params = model_params or {}

        self.models_: Dict[str, BaseForecaster] = {}
        self.weights_: Optional[np.ndarray] = None
        self.feature_names_: Optional[List[str]] = None

    def fit(
        self,
        X: pd.DataFrame,
        y: pd.Series,
        eval_set: Optional[tuple] = None,
        X_val: Optional[pd.DataFrame] = None,
        y_val: Optional[pd.Series] = None,
    ) -> 'ForecastingEnsemble':
        """
        Train all ensemble members.

        If weights='optimize', requires validation data to learn weights.
        """
        self.feature_names_ = list(X.columns)

        # Use eval_set if provided, otherwise use X_val, y_val
        if eval_set is not None:
            X_val, y_val = eval_set

        # Train each model
        for name in self.model_names:
            print(f"Training {name}...")
            model_class = MODEL_REGISTRY[name]
            params = self.model_params.get(name, {})

            model = model_class(**params)
            model.fit(X, y, eval_set=(X_val, y_val) if X_val is not None else None)
            self.models_[name] = model

        # Set weights
        if isinstance(self.weights_strategy, list):
            self.weights_ = np.array(self.weights_strategy)
        elif self.weights_strategy == 'equal':
            self.weights_ = np.ones(len(self.model_names)) / len(self.model_names)
        elif self.weights_strategy == 'optimize':
            if X_val is None:
                raise ValueError("Validation data required for weight optimization")
            self.weights_ = self._optimize_weights(X_val, y_val)

        print(f"Ensemble weights: {dict(zip(self.model_names, self.weights_))}")
        return self

    def _optimize_weights(
        self,
        X_val: pd.DataFrame,
        y_val: pd.Series,
    ) -> np.ndarray:
        """Learn optimal weights to minimize validation RMSE."""
        # Get predictions from each model
        predictions = np.column_stack([
            self.models_[name].predict(X_val)
            for name in self.model_names
        ])

        def objective(weights):
            weighted_pred = predictions @ weights
            return np.sqrt(np.mean((y_val.values - weighted_pred) ** 2))

        # Constraints: weights sum to 1, all non-negative
        constraints = {'type': 'eq', 'fun': lambda w: np.sum(w) - 1}
        bounds = [(0, 1) for _ in self.model_names]
        x0 = np.ones(len(self.model_names)) / len(self.model_names)

        result = minimize(
            objective, x0,
            method='SLSQP',
            bounds=bounds,
            constraints=constraints,
        )

        return result.x

    def predict(self, X: pd.DataFrame) -> np.ndarray:
        """Generate weighted ensemble predictions."""
        predictions = np.column_stack([
            self.models_[name].predict(X)
            for name in self.model_names
        ])
        return predictions @ self.weights_

    def predict_individual(self, X: pd.DataFrame) -> Dict[str, np.ndarray]:
        """Get predictions from each individual model."""
        return {
            name: model.predict(X)
            for name, model in self.models_.items()
        }

    def get_feature_importance(self) -> pd.DataFrame:
        """Get weighted average feature importance."""
        importance_dfs = []
        for name, model in self.models_.items():
            imp = model.get_feature_importance()
            imp['model'] = name
            importance_dfs.append(imp)

        combined = pd.concat(importance_dfs)
        avg_importance = combined.groupby('feature')['importance'].mean()

        return pd.DataFrame({
            'feature': avg_importance.index,
            'importance': avg_importance.values
        }).sort_values('importance', ascending=False)

    def get_params(self) -> Dict[str, Any]:
        return {
            'models': self.model_names,
            'weights': self.weights_strategy,
            'model_params': self.model_params,
        }

    def set_params(self, **params) -> 'ForecastingEnsemble':
        for key, value in params.items():
            if hasattr(self, key):
                setattr(self, key, value)
        return self
```

### 4. Training Script (`scripts/train_ensemble.py`)

```python
#!/usr/bin/env python
"""Train ensemble model."""
import argparse
from src.data.loader import load_sales_data
from src.features.pipeline import build_feature_pipeline
from src.models.ensemble import ForecastingEnsemble
from src.evaluation.splits import temporal_train_test_split
from src.evaluation.metrics import calculate_metrics, print_metrics


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--data", required=True)
    parser.add_argument("--weights", default="optimize", choices=["equal", "optimize"])
    args = parser.parse_args()

    # Load and prepare
    df = load_sales_data(args.data)
    pipeline = build_feature_pipeline()
    df_features = pipeline.fit_transform(df)

    feature_cols = [c for c in df_features.columns
                    if c not in ['date', 'sku_id', 'sales']]
    X = df_features[feature_cols].dropna()
    y = df_features.loc[X.index, 'sales']

    # Split: train / validation / test
    X_trainval, X_test, y_trainval, y_test = temporal_train_test_split(X, y, test_size=0.15)
    X_train, X_val, y_train, y_val = temporal_train_test_split(X_trainval, y_trainval, test_size=0.18)

    print(f"Train: {len(X_train):,}, Val: {len(X_val):,}, Test: {len(X_test):,}")

    # Train ensemble
    ensemble = ForecastingEnsemble(
        models=['lightgbm', 'xgboost', 'catboost'],
        weights=args.weights,
    )
    ensemble.fit(X_train, y_train, X_val=X_val, y_val=y_val)

    # Evaluate ensemble
    y_pred = ensemble.predict(X_test)
    metrics = calculate_metrics(y_test.values, y_pred)
    print("\nEnsemble Metrics:")
    print_metrics(metrics)

    # Compare to individual models
    print("\nIndividual Model Comparison:")
    for name, pred in ensemble.predict_individual(X_test).items():
        m = calculate_metrics(y_test.values, pred)
        print(f"  {name}: MAPE={m['mape']:.2%}, RMSE={m['rmse']:.2f}")


if __name__ == "__main__":
    main()
```

## Implementation Checklist

- [ ] Implement `XGBoostForecaster` with consistent interface
- [ ] Implement `CatBoostForecaster` with native categorical support
- [ ] Create `ForecastingEnsemble` with weight optimization
- [ ] Add individual model prediction access
- [ ] Implement weighted feature importance
- [ ] Handle categorical features correctly per model
- [ ] Create training script with comparison output

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/models/xgboost_model.py` | 0-120 | Create | XGBoost wrapper |
| `src/models/catboost_model.py` | 0-120 | Create | CatBoost wrapper |
| `src/models/ensemble.py` | 0-180 | Create | Ensemble orchestration |
| `scripts/train_ensemble.py` | 0-60 | Create | Training script |

## Technical Challenges

**Categorical Feature Handling**: Each library handles categoricals differently.

```python
# LightGBM: Label encode or use native
lgb_train = lgb.Dataset(X, label=y, categorical_feature=['category'])

# XGBoost: Requires numeric encoding
X['category_encoded'] = LabelEncoder().fit_transform(X['category'])

# CatBoost: Handles strings natively
model.fit(X, y, cat_features=['category'])
```

**Solution**: Create preprocessing per model type.

## Definition of Done

- [ ] All three models train successfully
- [ ] Weight optimization converges
- [ ] Ensemble outperforms best individual model
- [ ] Feature importance aggregates correctly
- [ ] Tests pass: `pytest tests/test_ensemble.py -v`

## Risk Assessment

| Risk | Impact | Mitigation |
|------|--------|------------|
| CatBoost categorical encoding errors | :yellow_circle: Medium | Auto-detect and handle |
| Weight optimization failing | :yellow_circle: Medium | Fall back to equal weights |
| Memory issues with 3 models | :green_circle: Low | Train sequentially |

## Time Estimate

**Estimated Hours**: 8-12 hours
**Priority**: Medium (improves accuracy)
